doc($sce): Overhaul the $sce service's documentation.

rjamet opened the pull request:
A big docs update around the $sce: there's a lot of content in there that's often misunderstood, and some of the documentation starts to get really old too. Plus some minor falsehoods in there... Overall, this isn't a major refactoring, but there's a lot of smaller changes.

cc @mprobst for another set of eyes on the contents.

**Does this PR introduce a breaking change?**
Nope, it's a doc-only CL.

**Other information**:
- I haven't tested how the doc renders, as I couldn't find a way to do so ?

rjamet committed:
doc($sce): Overhaul the $sce service's documentation.

A big docs update around the $sce: there's a lot of content in there that's often misunderstood, and some of the documentation starts to get really old too.

mprobst commented on src/ng/sce.js:
```diff
@@ -855,11 +921,9 @@ function $SceProvider() {
-     *     return value of {@link ng.$sce#trustAs $sce.trustAs}.)
+     * @param {*} value The value to mark as trusted for URL context.
+     * @returns {*} A wrapped version of value that can be used as a trusted variant
+     *   of your `value` in URL context (currently unused.)
```
(currently untrusted) is a bit misleading, maybe `used as a trusted variant of your previously untrusted value`?

rjamet replied:
it's unused, not untrusted. Fixed by saying that

rjamet commented on src/ng/sce.js:
```diff
@@ -288,17 +322,23 @@ function $SceDelegateProvider() {
+     * for general documentation about strict contextual escaping.
+     *
+     * @param {string} type The kind of context in which this value is safe
+     *   for use.  e.g. url, resourceUrl, html, js and css.
```
Should I do that for the whole file, while I'm at it? Do you have a style guide for doc?

mprobst replied:
Please do. I don't think we have a style guide in Angular, but +4 matches Google JS style and what you generally do in other languages.

rjamet replied:
Ok, I'll do that. 80 or 100 cols, btw?

rjamet committed:
doc($sce): adress comments, plus more formatting fixes

rjamet commented on src/ng/sce.js:
```diff
@@ -288,17 +322,23 @@ function $SceDelegateProvider() {
+     * For instance, marking a string as trusted for the $sce.HTML context will
+     * entirely bypass the potential $sanitize call in corresponding $sce.HTML
+     * sinks, such as an ng-bind-html directive, while using the same value
+     * passed as a string will cause it to be sanitized.
```
Added "Note that in most cases you won't need to call trustAsHtml: if you have the sanitizer loaded, passing the value itself will render all HTML that does not pose a security risk."

rjamet commented on src/ng/sce.js:
```diff
@@ -288,17 +322,23 @@ function $SceDelegateProvider() {
-     * See {@link ng.$sce $sce} for enabling strict contextual escaping.
+     * Marks the parameter as trusted for the specified context, and returns the
+     * trusted representation of it. This trusted object should later on be
+     * used as-is, without any security check, by sinks that use this context.
```
I've replaced it by "used as is without any security checks by bindings or directives that require this security context."

rjamet commented on src/ng/sce.js:
```diff
@@ -288,17 +322,23 @@ function $SceDelegateProvider() {
-     * such as for onclick,  etc.) that uses the provided value.
-     * See {@link ng.$sce $sce} for enabling strict contextual escaping.
+     * Marks the parameter as trusted for the specified context, and returns the
+     * trusted representation of it. This trusted object should later on be
```
Yup, done.

rjamet commented on src/ng/sce.js:
```diff
@@ -288,17 +322,23 @@ function $SceDelegateProvider() {
-     * attribute interpolation, any dom event binding attribute interpolation
-     * such as for onclick,  etc.) that uses the provided value.
-     * See {@link ng.$sce $sce} for enabling strict contextual escaping.
+     * Marks the parameter as trusted for the specified context, and returns the
```
Yup, right, my formulation might imply that there's a global list of trusted stuff somewhere. Changed to "Returns a trusted representation of the parameter for the specified context. "

rjamet commented:
Some more formatting fixes after your comments:
- +4 spaces for indents
- 100 col everywhere
- s/@returns/@return/
- I uniformized context names : it's not "x for a kind of context" anymore, just "x for context", and always refer to them with the form `$sce.HTML` 
- a bunch of backticks here and there

mprobst commented:
LGTM.

gkalpak commented on src/ng/sce.js:
```diff
@@ -288,17 +319,24 @@ function $SceDelegateProvider() {
-     * @param {*} value The value that that should be considered trusted/safe.
-     * @returns {*} A value that can be used to stand in for the provided `value` in places
-     * where AngularJS expects a $sce.trustAs() return value.
+     * @param {*} value The value that that should be considered trusted.
```
_that that_ --> _that_

rjamet replied:
done

gkalpak commented on src/ng/sce.js:
```diff
@@ -351,33 +389,38 @@ function $SceDelegateProvider() {
-     * Disabling auto-escaping is extremely dangerous, it usually creates a Cross Site Scripting
-     * (XSS) vulnerability in your application.
-     * </div>
+     * In practice, there's several cases. When given a string, this run checks and sanitization to
```
_there's_ --> _there are_
_this run_ --> _it runs_

rjamet replied:
done

gkalpak commented on src/ng/sce.js:
```diff
@@ -351,33 +389,38 @@ function $SceDelegateProvider() {
-     * </div>
+     * In practice, there's several cases. When given a string, this run checks and sanitization to
+     * make it safe without prior assumptions. When given the result of a {@link
+     * ng.$sceDelegate#trustAs `$sceDelegate.trustAs`} call, this returns the originally supplied
```
_this_ --> _it_ (we generally try to avoid "this" as it can be confusing)

rjamet replied:
done (plus another this above clarified in this function)

gkalpak commented on src/ng/sce.js:
```diff
@@ -351,33 +389,38 @@ function $SceDelegateProvider() {
-     * @returns {*} The value the was originally provided to {@link ng.$sceDelegate#trustAs
-     *     `$sceDelegate.trustAs`} if valid in this context.  Otherwise, throws an exception.
+     *     `$sceDelegate.trustAs`} call, or anything else (which will not be considered trusted.)
+     * @return {*} A version of value that's safe to use in the given context, or throws an
```
_of value_ --> _of the value_

rjamet replied:
done

gkalpak commented on src/ng/sce.js:
```diff
@@ -423,21 +468,26 @@ function $SceDelegateProvider() {
- * to the top of your HTML document.
+ * To illustrate this, consider the `ng-bind-html` directive. It renders its value directly as HTML:
+ * we call that the *context*. When given an untrusted input, AngularJS will attempt to sanitize it
+ * before rendering if a sanitizer is available, and throw otherwise. To render the input as-is, you
```
To make this more obvious, I might prepend: _"To bypass sanitization and render the input as-in..."_

rjamet replied:
That's better, yes. fixed.

gkalpak commented on src/ng/sce.js:
```diff
@@ -460,25 +510,29 @@ function $SceDelegateProvider() {
+ * context.  That trust is formalized with a function call. This means that as a developer, you
+ * can assume all untrusted bindings are safe. Then, to audit your code for binding security issues,
+ * you just need to ensure the values you mark as trusted indeed are safe - because they were
+ * received from your server, sanitized by your library, etc.. You can organize your codebase to
```
_etc.._ --> _etc._

rjamet replied:
done

gkalpak commented on src/ng/sce.js:
```diff
@@ -519,11 +573,12 @@ function $SceDelegateProvider() {
- * through {@link ng.$sce#getTrusted $sce.getTrusted}.  SCE doesn't play a role here.
+ * call `$sce.trustAs` on them (e.g.
+ * `<div ng-bind-html="'<b>implicitly trusted</b>'"></div>`) just works. The `$sceDelegate` will
+ * also use the `$sanitize` injectable if it is available when binding untrusted values to
```
_injectable_ --> _service_ (It would be clearer to the reader, as this is the term we use throughout the docs.)

rjamet replied:
Yes, right

gkalpak commented on src/ng/sce.js:
```diff
@@ -840,11 +902,23 @@ function $SceProvider() {
+     *
+     * @param {*} value The value to mark as trusted for `$sce.CSS` context.
+     * @return {*} A wrapped version of value that can be used as a trusted variant
+     *     of your `value` in CSS context. This context is currently unused, so there are almost no
```
_CSS_ --> _\`$sce.CSS\`_ (?)

rjamet replied:
I missed that one. Fixed.

gkalpak commented on src/ng/sce.js:
```diff
@@ -898,16 +969,16 @@ function $SceProvider() {
+     * takes any input, and either returns a value that's safe to use in the specified context,
+     * or throws an exception. This function is aware of trusted values created by the `trustAs`
+     * function and its shorthands, and when contexts are appropriate, returns the unwrapped value
+     * as-is. Finally, this function can also throw when there is no way to turn maybeTrusted in a
```
_maybeTrusted_ --> _\`maybeTrusted\`_

rjamet replied:
Done

gkalpak commented on src/ng/sce.js:
```diff
@@ -898,16 +969,16 @@ function $SceProvider() {
-     *              Otherwise, throws an exception.
+     * @param {string} type The context in which this value is to be used.
+     * @param {*} maybeTrusted The result of a prior {@link ng.$sce#trustAs `$sce.trustAs`} call.
+     * @return {*} The value the was originally provided to {@link ng.$sce#trustAs `$sce.trustAs`}
```
_the was_ --> _that was_

rjamet replied:
done

gkalpak commented:
BTW, regarding the 4-space indentation on `@param` or `@return` descriptions: I think that is some cases (e.g. when there are bullet-points or empty lines) dgeni will recongnize this indended blocks as code blocks and wrap them in `<pre>` tags. Have you checked the rendered output?

rjamet committed:
doc($sce): Adress gkalpak's comments.

rjamet commented on src/ng/sce.js:
```diff
@@ -155,21 +185,22 @@ function $SceDelegateProvider() {
    * same origin resource requests.
    *
+   *     <div class="alert alert-warning">
+   *     **Note:** the default whitelist of 'self' is not recommanded if your app shares its origin
```
done

rjamet commented:
I don't know how to render these so far, so everything is blind edits... I'll look into dgeni.

gkalpak commented:
My only concern is whether the 4-space indentations make some blocks incorrectly rendered as code blocks. Other than that it LGTM.

@rjamet, you can see the rendered output with the folloing commands:

```
grunt package   # May take a while :)
grunt webserver
```

The open http://localhost:8000/build/docs/.

(After you have run `grunt package` once, you can run `grunt docs` (to only update the docs) and reload your browser to see the changes.)

rjamet commented:
Your fears are justified, that parses paragraphs in param/return blocks as code blocks if there's an empty line. 

I'm tempted to keep the +4 indentation, and work around this by either two spaces + simple line break (=`<br/>` in markdown), or `</p><p>`. Would that be fine? It lets us be consistent, even if it's not that pretty: the issue is then to know if anything else consumes this. BR loses paragraphs, paragraphs add markup that might not make sense in IDEs. 

I spotted a few unrelated issues otherwise, I'll bundle them as part of the next fix.

gkalpak commented:
Either way works for me, as long as the output looks the way it should.

rjamet committed:
doc($sce): more minor formatting/content fixes.

Some capitalization was wrong, formatting fixes for +4 indentation on @param tags,
and uniformize some params descriptions on the same functions (scedelegate / sce).

rjamet commented:
Ok, fixed:
- capitalization of the return sections 
- some sce/scedelegate uniformization between getTrusted parameters 
- one "js" that should have been `$sce.JS`
- default values for the resource url white/blacklists should be described in description, not in return 
- +4 indentation everywhere it makes sense, `</p><p>` to keep paragraphs in there, except on the parseAs functions where I kept the +3 because +4-indented lists resist any dirty hack I could find :/

I checked all the $sce doc output, it should be the same where I didn't change the text.

rjamet committed:
doc($sce): add missing paragraph in resourceUrlBlacklist

gkalpak reviewed (approved):
I trust you that it looks good when rendered :smiley:
(You still need to fix the linting issue Travis complains about.)

rjamet committed:
doc($sce): trailing space

rjamet commented:
Oops, sorry, I missed it. Fixed.
